esp-hal/src/gpio/asynch.rs

```diff
@@ -0,0 +1,223 @@
+    /// The GPIO driver will disable listening for the event once it occurs,
+    /// or if the `Future` is dropped - which also means this method is **not**
+    /// cancellation-safe, it will always wait for a future event.
+    ///
+    /// Note that calling this function will overwrite previous
+    /// [`listen`][Self::listen] operations for this pin.
+    #[inline]
+    pub async fn wait_for(&mut self, event: Event) {
+        self.pin.wait_for(event).await
+    }
+
+    /// Wait until the pin is high.
+    ///
+    /// See [Self::wait_for] for more information.
+    #[inline]
+    pub async fn wait_for_high(&mut self) {
+        self.pin.wait_for_high().await
+    }
+
+    /// Wait until the pin is low.
+    ///
+    /// See [Self::wait_for] for more information.
+    #[inline]
+    pub async fn wait_for_low(&mut self) {
+        self.pin.wait_for_low().await
+    }
+
+    /// Wait for the pin to undergo a transition from low to high.
+    ///
+    /// See [Self::wait_for] for more information.
+    #[inline]
+    pub async fn wait_for_rising_edge(&mut self) {
+        self.pin.wait_for_rising_edge().await
+    }
+
+    /// Wait for the pin to undergo a transition from high to low.
+    ///
+    /// See [Self::wait_for] for more information.
+    #[inline]
+    pub async fn wait_for_falling_edge(&mut self) {
+        self.pin.wait_for_falling_edge().await
+    }
+
+    /// Wait for the pin to undergo any transition, i.e low to high OR high
+    /// to low.
+    ///
+    /// See [Self::wait_for] for more information.
+    #[inline]
+    pub async fn wait_for_any_edge(&mut self) {
+        self.pin.wait_for_any_edge().await
+    }
+}
+
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+struct PinFuture<'f, 'd> {
+    pin: &'f mut Flex<'d>,
+}
+
+impl PinFuture<'_, '_> {
+    fn number(&self) -> u8 {
+        self.pin.number()
+    }
+
+    fn bank(&self) -> GpioBank {
+        self.pin.pin.bank()
+    }
+
+    fn mask(&self) -> u32 {
+        self.pin.pin.mask()
+    }
+
+    fn is_done(&self) -> bool {
+        // Only the interrupt handler should clear the async bit, and only if the
+        // specific pin is handling an interrupt. This way the user may clear the
+        // interrupt status without worrying about the async bit being cleared.
+        self.bank().async_operations().load(Ordering::Acquire) & self.mask() == 0
+    }
+}
+
+impl core::future::Future for PinFuture<'_, '_> {
+    type Output = ();
+
+    fn poll(mut self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        PIN_WAKERS[self.number() as usize].register(cx.waker());
+
+        if self.is_done() {
+            self.pin.clear_interrupt();
+
+            // Forget self - we don't want to take a critical section to deconfigure the
+            // interrupt, and we don't need to - it's done by the interrupt
+            // handler.
+            #[allow(clippy::forget_non_drop)]
+            core::mem::forget(self);
```

Comment:
I think this doesn't do anything. `Drop` will still be called, as only a mere reference has been forgotten.
Hence the `#[allow(clippy::forget_non_drop)]`.